### PR TITLE
[DD4hep] Tool for creation of DD4hep geometry DB payloads and for migration validation

### PR DESCRIPTION
**Pull request by cvuosalo**
There is a tool for expert use to create geometry DB payloads from DDD XML files. This PR provides a DD4hep version of the same tool. An initial decision was made to match the design and function of the DDD tool to reduce development risks and to facilitate validation.

Note that this tool is intended for the creation of Run 3 DD4hep geometry DB payloads. At this time, it is not anticipated that Runs 1-2 DB geometry payloads will be re-built. For Phase 2, the suitability of this design will have to be assessed, and changes may be necessary.

This tool produces the complete detector geometry in a human-readable XML file, which is not only suitable for loading into the DB but also can be used for diagnostics and validation.

There are two minor issues that will be addressed in follow-up PRs. The DB payload creation script currently has the reduced material scenarios commented out. A subsequent PR will implement those scenarios. Secondly, PPS uses a special XML payload for reconstruction, which is currently created with DDD. A later PR will implement creation of this payload with DD4hep.

Validation
The extended geometry description DB payload created by the tool was validated against its DDD counterpart incrementally, section by section, to ensure the content was identical. A stress test where the tool read its own output was also done to verify it would reproduce that output. Also, 100 ttbar events generated with DD4hep XML and DD4hep DB payloads were checked to verify that there were no significant differences in results. A RelVal is underway for further validation.

Reco geometry DB payloads were checked to see that they are identical to the ones created with DDD. The Tracker Reco geometry DB payload created with DD4hep is not numerically identical to the DDD version, but it was thoroughly checked to see that the content is the same as DDD, except for insignificant numerical variations. The output of `DDDCmsTrackerContruction::printAllTrackerGeometricDets` was checked, as well as the `ModuleInfo.log`, `ModuleNumbering.dat`, `TECLayout_CMSSW.dat` outputs from `Geometry/TrackerGeometryBuilder/test/ModuleInfo.cc`, and all were found to be the same for DD4hep and DDD. The only differences were phi angles equal to |pi| that varied between -pi and +pi, and tiny values near 0 that could varied between -1e-12 and +1e-12.

To avoid these variations in future comparisons when new DB payloads are candidates for uploading, this PR proposes that values in the DD4hep Tracker Reco geometry record (the `PGeometricDet`) be standardized as follows:
* phi values of |pi| will be set to +pi.
* Translation x, y, z values and rotation matrix elements that are <=|1e-10| will be set to true 0.
This proposal has already been discussed with the Tracker DPG.

### CODE REVIEW

File: DetectorDescription/OfflineDBLoader/plugins/OutputDDToDDL.cc
Comment (cvuosalo) [inline DetectorDescription/OfflineDBLoader/plugins/OutputDDToDDL.cc:136]: This file will not be part of the final PR. It was only edited for debugging purposes.

File: DetectorDescription/DDCMS/src/DDDetector.cc
Comment (ianna) [inline DetectorDescription/DDCMS/src/DDDetector.cc:23]: @cvuosalo - is it a memory leak?

**Review by ianna — CHANGES_REQUESTED** [review]
@cvuosalo - IMHO, the whole approach is seriously flawed. It tries to re-implement the DDD [Stores](https://cmssdt.cern.ch/lxr/source/DetectorDescription/Core/interface/Store.h), but does it half way. It uses a full Geant4 geometry where a simpler and more elegant approach would do (e.g. DDD [Graph](https://cmssdt.cern.ch/lxr/source/DataFormats/Math/interface/Graph.h)). 

@makortel - FYI

**Comment by cvuosalo** [comment]
Refinement and simplification is certainly possible as time allows. My first priority was to try to match the old DD DB payload as closely as possible. This whole design will need to be revisited for Phase 2 DD4hep, which has more challenging requirements than Run 3. I think Phase 2 is a good opportunity for considering new approaches for the DB payloads, since the DB payloads will be completely new.  Some of the code in this PR might be useful for Phase 2, or maybe it will be dropped for an entirely different method.

File: DetectorDescription/DDCMS/src/DDDetector.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/src/DDDetector.cc:23]: See below on line 34. In general use, when this code is not creating a DB payload, the parsing context is removed after the end of parsing. If I understand DD4hep correctly, the `removeExtension` method issues a delete on the pointer that was passed to `addExtension`.

**Comment by ianna** [comment]
> Refinement and simplification is certainly possible as time allows. My first priority was to try to match the old DD DB payload as closely as possible. This whole design will need to be revisited for Phase 2 DD4hep, which has more challenging requirements than Run 3. I think Phase 2 is a good opportunity for considering new approaches for the DB payloads, since the DB payloads will be completely new. Some of the code in this PR might be useful for Phase 2, or maybe it will be dropped for an entirely different method.

The problem with this temporary solution is that an `Assembly` as an XML description is ignored. Once a payload is produced using this approach, we’d have to support it. This has a long time implication and restricts the future design. I’d suggest we do not use a DB payload for **an ideal geometry description** in DD4hep-based workflows until we have a proper solution.

File: DetectorDescription/DDCMS/interface/DDXMLTags.h
Comment (ianna) [inline DetectorDescription/DDCMS/interface/DDXMLTags.h:140]: @cvuosalo - why is it needed?

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (ianna) [inline DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:868]: @cvuosalo - why not use <ShapelessSolid> to (re-)store an Assembly?

**Review by ianna — CHANGES_REQUESTED** [review]
some more questions

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:868]: Yes, this tag was my workaround since Assembly had no definition in XML until yesterday. But now that we can define assemblies in XML, this workaround can be dropped and replaced with explicit Assembly definitions, which are much more elegant.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (ianna) [inline DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:971]: x@cvuosalo - same here, please, let dd4hep handle it.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (ianna) [inline DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:1723]: @cvuosalo - same here, please, let dd4hep handle it. Assembly is a shapeless solid.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (ianna) [inline DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:2166]: @cvuosalo - please, revert.

File: DetectorDescription/DDCMS/src/DDDetector.cc
Comment (ianna) [reply DetectorDescription/DDCMS/src/DDDetector.cc:23]: Please, do not expose an XML parsing context beyond an XML parser. Thanks.

File: DetectorDescription/DDCMS/src/DDFilteredView.cc
Comment (ianna) [inline DetectorDescription/DDCMS/src/DDFilteredView.cc:556]: @cvuosalo - it should return `cms::DDSolidShape::ddshapelesssolid`

File: DetectorDescription/DDCMS/src/DDNamespace.cc
Comment (ianna) [inline DetectorDescription/DDCMS/src/DDNamespace.cc:269]: @cvuosalo - please, use `ns.volume(name)`

File: DetectorDescription/OfflineDBLoader/BuildFile.xml
Comment (ianna) [inline DetectorDescription/OfflineDBLoader/BuildFile.xml:6]: @cvuosalo - Introducing G4 dependency adds an extra layer of conversion. That means we will not have the same geometry description between the first XML parsing from the files and the second XML parsing from the DB payload.

**Review by ianna — CHANGES_REQUESTED** [review]
@cvuosalo - I see it is still a draft. Please, do not expose the parsing context via a `DDDetectror`. Thanks.

File: DetectorDescription/OfflineDBLoader/BuildFile.xml
Comment (cvuosalo) [reply DetectorDescription/OfflineDBLoader/BuildFile.xml:6]: The G4 dependency is a left-over. It will be removed soon.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:2166]: `OutputDD4hepToDDL.cc` needs the parsing context. This approach allows the parsing context to be passed to `OutputDD4hepToDDL.cc`. At the moment, I don't see a good alternative that can be implemented on a reasonable timescale.

File: DetectorDescription/DDCMS/src/DDFilteredView.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/src/DDFilteredView.cc:556]: That would be better than a Box. This change will be made shortly.

File: DetectorDescription/DDCMS/interface/DDDetector.h
Comment (cvuosalo) [reply DetectorDescription/DDCMS/interface/DDDetector.h:39]: Maybe it should be a method.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:873]: I tried that, and I couldn't get it to work. In a `PosPart` secton, when one of the volumes is an assembly, the assembly has to be treated as an assembly and not a volume. That is why I use polymorphic pointers.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:1723]: Line 1723 is bookkeeping to keep track of all the Assembly names in the Solid Section. Adding the Box is probably not necessary, but I put it in to be safe.

File: DetectorDescription/DDCMS/src/DDNamespace.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/src/DDNamespace.cc:269]: A polymorphic pointer is needed. `volumes` seems to contain pure volumes.

File: DetectorDescription/DDCMS/src/DDFilteredView.cc
Comment (cvuosalo) [reply DetectorDescription/DDCMS/src/DDFilteredView.cc]: Actually, a Box is needed for consistency and to pass the DDSolid.cppunit.cc unit test.

**Comment by cvuosalo** [comment]
I discovered an issue where some DDD components were included when using DD4hep to build the Calo Reco payloads. Sunanda and I are working on it. Otherwise, this PR is almost ready for final review.

**Comment by cvuosalo** [comment]
The Calo Reco payload issue is solved, but now I have found that the Tracker Reco geometry has a small issue when built with DD4hep. `Geometry/​TrackerNumberingBuilder/​src/​GeometricDet.cc` may need a small revision.

**Comment by cvuosalo** [comment]
@ianna Changes have been made in this PR to try to address concerns.
For `DDDetector`, the parsing context member has been removed, as has the explicit "makePayloads" parameter.

`DDDetectorESProducer` has been returned to its original form, with no defined parameter for payloads.

The map of volume pointers has been deleted from `DDNamespace`, and the code for handling assemblies and volumes has thereby been somewhat simplified. However, I could not figure out how to simplify Assembly processing any further. When the DD4hep extended geometry description in the DB is being read, which contains explicit Assembly definitions, the code has to treat Assemblies differently from plain Volumes. In my initial implementation attempt, I tried to treat Assemblies just like other Volumes, but I could not get that to work. The only way I could get Assembly processing to work was to have the code explicitly check whether a volume was actually an Assembly and treat it differently if it was.

I'm not sure what other issues might remain with this PR. While trying to wrap up, I discovered issues with the creation of the Calo and Tracker Reco geometry records that took a lot of time to investigate and resolve.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@adewit @mmusich Please see the changes that we discussed in `Geometry/TrackerNumberingBuilder/src/GeometricDet.cc` and ` CondTools/Geometry/plugins/PGeometricDetBuilder.cc`.

**Comment by cvuosalo** [comment]
This PR is aimed for 12_0_0_pre5.

**Comment by cvuosalo** [comment]
The unit test errors appear to be unrelated to this PR and seem to be in the IB and related to:
#34601
#34587

**Comment by cvuosalo** [comment]
The Event Setup warnings in `CondTools/Geometry/plugins/PGeometricDetBuilder.cc` should be resolved in a separate PR. The DB group may already be working on this as part of the ES migration to tokens.

**Comment by cvuosalo** [comment]
@cms-sw/db-l2 Could you please review this PR soon? I would like to try to include it in 12_0_0_pre5 that is due July 27th. Thanks.

**Comment by cvuosalo** [comment]
@cmsbuild please test with #34622

**Comment by civanch** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by civanch** [comment]
urgent

File: CondTools/Geometry/test/writehelpers/createExtended2021DD4hepPayloads.sh
Comment (ianna) [inline CondTools/Geometry/test/writehelpers/createExtended2021DD4hepPayloads.sh:30]: @cvuosalo - please, remove commented out code.

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc
Comment (ianna) [inline Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:118]: @cvuosalo - how much does it cost in terms of performance loss?

**Review by ianna — CHANGES_REQUESTED** [review]
@cvuosalo - I still have a number of concerns about this code. We should discuss it offline.

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc
Comment (civanch) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:118]: @ianna , I agree with your concerns but just today we should decide to merge this PR or to close it. We cannot improve it anymore. 

I would propose to merge and  for the  next release 12_1_0 we should put efforts for the code optimization.

File: CondTools/Geometry/test/writehelpers/createExtended2021DD4hepPayloads.sh
Comment (cvuosalo) [reply CondTools/Geometry/test/writehelpers/createExtended2021DD4hepPayloads.sh:30]: As mentioned in the PR description, a forthcoming PR will enable these lines to produce the reduced material scenarios. Right now the lines serve as reminders of what will be implemented soon.

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc
Comment (cvuosalo) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:118]: These lines are executed when the `GeometricDet` is being built with DD4hep from XML. In production, the `GeometricDet` is read from the `PGeometricDet` in the DB, so this constructor is not called.
As a quick check, I looked at the times for the 11634.911 DD4hep workflow that reads geometry from XML. In the PR test for this PR (with the change) and the PR test for recent PR #34625, the average time per event in step1 simulation was 23 seconds in both cases, so there appears to be no large effect. Tests with larger numbers of events would be needed to see if there is a measurable effect.

Comment (cvuosalo) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:118]: A quick check of the start-up time (including XML processing) of DD4hep XML workflow 11634.911 shows no significant difference between this PR and another PR test without this PR. More careful checks will be needed to determine whether this added function has a measurable effect on performance. Note that the `almostEqual` function is a simple template, so it shouldn't have much overhead.

**Comment by qliphy** [comment]
+1

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc
Comment (ianna) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:118]: > These lines are executed when the GeometricDet is being built with DD4hep from XML. In production, the GeometricDet is read from the PGeometricDet in the DB, so this constructor is not called.

This is not true for the Phase 2 geometry that is using XML as a main source.

Comment (ianna) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:118]: > More careful checks will be needed to determine whether this added function has a measurable effect on performance. Note that the almostEqual function is a simple template, so it shouldn't have much overhead.

Yes, careful performance measurement ought to be done before merging the code. Any operation times the number of `GeometricDets` will have an effect.

**Pull request merged**
